src/frequenz/sdk/microgrid/component_data.py

```diff
@@ -17,229 +19,218 @@
 from .component_states import EVChargerCableState
 
 
-class _BaseComponentData(ABC):
+@dataclass(frozen=True)
+class ComponentData(ABC):
     """A private base class for strongly typed component data classes."""
 
-    _raw: microgrid_pb.ComponentData
+    component_id: int
+    timestamp: datetime
 
-    @property
-    def component_id(self) -> int:
-        """Return the component id of the message.
+    @staticmethod
+    def from_proto(raw: microgrid_pb.ComponentData) -> ComponentData:
+        """Create ComponentData from a protobuf message.
 
-        Returns:
-            Component id of the message.
+        Args:
+            raw: protobuf message
         """
-        return self._raw.id
-
-    @property
-    def timestamp(self) -> datetime:
-        """Return the timestamp of the message.
 
-        Returns:
-            Timestamp of the message.
-        """
-        return self._raw.ts.ToDatetime(tzinfo=pytz.UTC)
 
+@dataclass(frozen=True)
+class MeterData(ComponentData):
+    """A wrapper class for holding meter data.
 
-class MeterData(_BaseComponentData):
-    """A wrapper class for holding meter data."""
+    component_id: the component ID
```

Comment:
The diff makes it really hard to read, but the Docstring format is missing an `Attributes:` header:

```py
class MeterData(ComponentData):
    """A wrapper class for holding meter data.

    Attributes:
        component_id: the component ID
        ...
```

Not sure why this isn't caught by darglint or the other.

https://google.github.io/styleguide/pyguide.html#384-classes

Reply:
Because the `Attributes` section isn't enforced by our checkers. Just look at other modules in the SDK—it is not there.

I didn't even know there's `Attributes` section, I just copied the docstrings from protobuf messages in the following format `attr: docstring`.

Reply:
Oh, that's quite crappy, the style guide clearly allows documenting attributes that are not part of the the `__init__()`, it is in the example.

BTW, did you really tried this? Because I just did and it worked for me (based on the old version):

```diff
diff --git a/src/frequenz/sdk/microgrid/component_data.py b/src/frequenz/sdk/microgrid/component_data.py
index ffacc35..7d94e9f 100644
--- a/src/frequenz/sdk/microgrid/component_data.py
+++ b/src/frequenz/sdk/microgrid/component_data.py
@@ -18,7 +18,12 @@ from .component_states import EVChargerCableState
 
 
 class _BaseComponentData(ABC):
-    """A private base class for strongly typed component data classes."""
+    """A private base class for strongly typed component data classes.
+
+    Attributes:
+        component_id: The component ID
+
+    """
 
     _raw: microgrid_pb.ComponentData
 
```

```console
$ nox -s docstrings
nox > Running session docstrings
nox > Creating virtual environment (virtualenv) using python3.8 in .nox/docstrings
nox > python -m pip install pydocstyle darglint toml
nox > pydocstyle src tests examples benchmarks
nox > darglint src examples
nox > Session docstrings was successful.
```

Reply:
You're right. It was `PyCharm` that was reporting a warning, but `darglint` works. Will add the `Attributes:` section.

<img width="758" alt="image" src="https://user-images.githubusercontent.com/109511717/194032381-634cec71-dae1-42cb-9e87-7c27c4095cda.png">

Reply:
Cool. We'll see when we render the docs what happens...